From b2ffecc5f6290d569cc354a5e72e0674255e6a0c Mon Sep 17 00:00:00 2001
From: Mithun RK <mithunr@yahoo-inc.com>
Date: Mon, 31 Jul 2017 14:54:02 -0700
Subject: [PATCH 1279/1363] CDH-57894 HIVE-17218: Canonical-ize hostnames for
 Hive metastore, and HS2 servers.

Conflicts:
	jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java

Change-Id: I49f0a2ad4c7c94e3e59117d9b62d571722db4813
---
 .../java/org/apache/hive/jdbc/HiveConnection.java  |    3 ++-
 .../hadoop/hive/metastore/HiveMetaStoreClient.java |   10 +++++++++-
 .../hadoop/hive/thrift/HadoopThriftAuthBridge.java |   16 ++++++++++++++++
 3 files changed, 27 insertions(+), 2 deletions(-)

diff --git a/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java b/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java
index 1d5d785..e2cf752 100644
--- a/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java
+++ b/jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java
@@ -19,6 +19,7 @@
 package org.apache.hive.jdbc;
 
 import org.apache.hadoop.hive.common.auth.HiveAuthUtils;
+import org.apache.hadoop.hive.shims.ShimLoader;
 import org.apache.hive.jdbc.Utils.JdbcConnectionParams;
 import org.apache.hive.service.auth.HiveAuthFactory;
 import org.apache.hive.service.auth.KerberosSaslHelper;
@@ -129,7 +130,7 @@ public HiveConnection(String uri, Properties info) throws SQLException {
     // sess_var_list -> sessConfMap
     // hive_conf_list -> hiveConfMap
     // hive_var_list -> hiveVarMap
-    host = connParams.getHost();
+    host = ShimLoader.getHadoopThriftAuthBridge().getCanonicalHostName(connParams.getHost());
     port = connParams.getPort();
     sessConfMap = connParams.getSessionVars();
     hiveConfMap = connParams.getHiveConfs();
diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
index a770b0b..4316682 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java
@@ -229,7 +229,15 @@ public HiveMetaStoreClient(HiveConf conf, HiveMetaHookLoader hookLoader)
             throw new IllegalArgumentException("URI: " + s
                 + " does not have a scheme");
           }
-          metastoreUris[i++] = tmpUri;
+          metastoreUris[i++] = new URI(
+              tmpUri.getScheme(),
+              tmpUri.getUserInfo(),
+              ShimLoader.getHadoopThriftAuthBridge().getCanonicalHostName(tmpUri.getHost()),
+              tmpUri.getPort(),
+              tmpUri.getPath(),
+              tmpUri.getQuery(),
+              tmpUri.getFragment()
+          );
 
         }
         // make metastore URIS random
diff --git a/shims/common/src/main/java/org/apache/hadoop/hive/thrift/HadoopThriftAuthBridge.java b/shims/common/src/main/java/org/apache/hadoop/hive/thrift/HadoopThriftAuthBridge.java
index 2e80171..4956087 100644
--- a/shims/common/src/main/java/org/apache/hadoop/hive/thrift/HadoopThriftAuthBridge.java
+++ b/shims/common/src/main/java/org/apache/hadoop/hive/thrift/HadoopThriftAuthBridge.java
@@ -22,6 +22,7 @@
 import java.io.IOException;
 import java.net.InetAddress;
 import java.net.Socket;
+import java.net.UnknownHostException;
 import java.security.PrivilegedAction;
 import java.security.PrivilegedExceptionAction;
 import java.util.Locale;
@@ -115,6 +116,21 @@ public String getServerPrincipal(String principalConfig, String host)
     return serverPrincipal;
   }
 
+  /**
+   * Method to get canonical-ized hostname, given a hostname (possibly a CNAME).
+   * This should allow for service-principals to use simplified CNAMEs.
+   * @param hostName The hostname to be canonical-ized.
+   * @return Given a CNAME, the canonical-ized hostname is returned. If not found, the original hostname is returned.
+   */
+  public String getCanonicalHostName(String hostName) {
+    try {
+      return InetAddress.getByName(hostName).getCanonicalHostName();
+    }
+    catch(UnknownHostException exception) {
+      LOG.warn("Could not retrieve canonical hostname for " + hostName, exception);
+      return hostName;
+    }
+  }
 
   public UserGroupInformation getCurrentUGIWithConf(String authMethod)
       throws IOException {
-- 
1.7.9.5

